dvc/fs/base.py

```diff
@@ -317,25 +317,19 @@ def _download_dir(
         download_files = FsspecCallback.wrap_fn(callback, self._download_file)
         max_workers = jobs or self.jobs
         with ThreadPoolExecutor(max_workers=max_workers) as executor:
-            futures = [
-                executor.submit(download_files, from_info, to_info)
-                for from_info, to_info in zip(from_infos, to_infos)
-            ]
-
             # NOTE: unlike pulling/fetching cache, where we need to
             # download everything we can, not raising an error here might
             # turn very ugly, as the user might think that he has
             # downloaded a complete directory, while having a partial one,
             # which might cause unexpected results in his pipeline.
-            for future in as_completed(futures):
-                # NOTE: executor won't let us raise until all futures that
-                # it has are finished, so we need to cancel them ourselves
-                # before re-raising.
-                exc = future.exception()
-                if exc:
-                    for entry in futures:
-                        entry.cancel()
-                    raise exc
+            try:
+                for _ in executor.imap_unordered(
+                    download_files, from_infos, to_infos
+                ):
+                    pass
+            except Exception:
+                executor.shutdown(wait=False, cancel_futures=True)
```

Comment:
[`cancel_futures`](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor.shutdown) is only available since Python3.9.

Reply:
changed this to allow cancelling unfinished tasks when en exception is raised in our `imap_unordered` implementation